tools/c7n_gcp/c7n_gcp/actions/labels.py

```diff
@@ -0,0 +1,258 @@
+# Copyright 2019 Karol Lassak
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import datetime, timedelta
+from dateutil import tz as tzutil
+
+from c7n.utils import type_schema
+from c7n.filters import FilterValidationError
+from c7n.filters.offhours import Time
+from c7n.lookup import Lookup
+from c7n_gcp.actions import MethodAction
+from c7n_gcp.filters.labels import LabelActionFilter
+
+from c7n_gcp.provider import resources as gcp_resources
+
+
+class BaseLabelAction(MethodAction):
+
+    def get_labels_to_add(self, resource):
+        return None
+
+    def get_labels_to_delete(self, resource):
+        return None
+
+    def _merge_labels(self, current_labels, new_labels, remove_labels):
+        result = current_labels
```

Comment:
minor I'd suggest making a shallow copy here ala result = dict(current_labels) since to avoid mutations during actions of in memory state.